crates/harp/src/srcref.rs

```diff
@@ -0,0 +1,152 @@
+//
+// srcref.rs
+//
+// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+//
+//
+
+use anyhow::anyhow;
+use libr::SEXP;
+use stdext::unwrap;
+
+use crate::exec::RFunction;
+use crate::exec::RFunctionExt;
+use crate::vector::IntegerVector;
+use crate::vector::Vector;
+use crate::RObject;
+
+/// Structured representation of `srcref` integer vectors
+/// 0-based offsets.
+#[derive(Debug)]
+pub struct SrcRef {
+    /// Lines and parsed lines may differ if a `#line` directive is used in code:
+    /// the former will respect the directive, the latter will just count lines.
+    pub line: std::ops::Range<usize>,
+    pub line_parsed: std::ops::Range<usize>,
+
+    /// Bytes and columns may be different due to multibyte characters.
+    pub column: std::ops::Range<usize>,
+    pub column_byte: std::ops::Range<usize>,
+}
+
+// Takes user-facing object as input. The srcrefs are retrieved from
+// attributes.
+impl RObject {
+    pub fn srcrefs(&self) -> anyhow::Result<Vec<SrcRef>> {
+        let srcref = unwrap!(self.attr("srcref"), None => {
+            return Err(anyhow!("Can't find `srcref` attribute"));
+        });
+
+        unsafe {
+            crate::List::new(srcref.sexp)?
+                .iter()
+                .map(|x| SrcRef::try_from(RObject::view(x)))
+                .collect()
+        }
+    }
+}
+
+// Takes individual `srcref` attribute as input
+impl TryFrom<RObject> for SrcRef {
+    type Error = anyhow::Error;
+
+    fn try_from(value: RObject) -> anyhow::Result<Self> {
+        crate::r_assert_type(value.sexp, &[libr::INTSXP])?;
+        crate::r_assert_capacity(value.sexp, 6)?;
+
+        let value = unsafe { IntegerVector::new(value)? };
+
+        let line = std::ops::Range {
+            start: (value.get_value(0)? - 1) as usize,
+            end: (value.get_value(2)? - 1) as usize,
+        };
+
+        let line_parsed = if unsafe { value.len() >= 8 } {
+            std::ops::Range {
+                start: (value.get_value(6)? - 1) as usize,
+                end: (value.get_value(7)? - 1) as usize,
+            }
+        } else {
+            line.clone()
+        };
+
+        let column = std::ops::Range {
+            start: (value.get_value(4)? - 1) as usize,
+            end: value.get_value(5)? as usize,
+        };
+
+        let column_byte = std::ops::Range {
+            start: (value.get_value(1)? - 1) as usize,
+            end: value.get_value(3)? as usize,
+        };
```

Comment:
Does `end: value.get_value(5)?` and `end: value.get_value(3)?` need a `-1`? 

If not, I think we need some comments in here about the adjustments made to the structure of the ranges, and what we expect the result to look like, i.e. is the resulting range `[ ]` or `[ )` ?

Reply:
The resulting range is the conventional `[ )`